File: day10/haskell/src/Main.hs
module Main (main) where

main :: IO ()
main = do
  input <- readFile "../input"
  let ss = reverse $ states input
  putStrLn $ "First answer: " ++ firstAnswer ss
  putStrLn $ "Second answer:\n" ++ secondAnswer ss
  where
    firstAnswer = show . sum . map (uncurry (*)) . take 6 . everyNth 40 . drop 19 . zip [1..]
    secondAnswer = unlines . chunksOf 40 . map pixel . zip screenCoords

states :: String -> [Int]
states = snd . foldl exec (1, []) . map words . lines
  where
    exec (x, ss) ["noop"]    = (x, x:ss)
    exec (x, ss) ["addx", n] = (x + read n, x:x:ss)
    exec _ _ = undefined

screenCoords :: [Int]
screenCoords = map (flip mod 40) [0..]

pixel :: (Int, Int) -> Char
pixel (screenX, x) = if x - 1 <= screenX && screenX <= x + 1 then '#' else '.'

everyNth :: Int -> [a] -> [a]
everyNth n = map snd . filter ((== 0) . flip mod n . fst) . zip [0..]

chunksOf :: Int -> [a] -> [[a]]
chunksOf _ [] = []
chunksOf n xs = take n xs : chunksOf n (drop n xs)

File: day10/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../input").expect("input file missing");
    let mut x = 1;
    let mut states = Vec::new();
    for line in input.lines() {
        let parts: Vec<&str> = line.split(' ').collect();
        match parts[0] {
            "addx" => {
                let n = parts[1].parse::<i32>().unwrap();
                states.push(x);
                states.push(x);
                x += n;
            }
            "noop" => {
                states.push(x);
            }
            _ => unreachable!(),
        }
    }
    let first_answer = states.iter()
        .enumerate()
        .skip(19)
        .step_by(40)
        .take(6)
        .map(|(cycle, x)| (1 + cycle as i32) * *x)
        .sum::<i32>();
    println!("First answer: {first_answer}");

    let width = 40;
    let height = 6;
    let mut screen = vec![vec!['.'; width]; height];
    for (cycle, x) in states.iter().enumerate() {
        let screen_x = cycle % width;
        let screen_y = cycle / width;
        if x - 1 <= screen_x as i32 && screen_x as i32 <= x + 1 {
            screen[screen_y][screen_x] = '#';
        }
    }
    println!("Second answer:");
    for line in screen {
        for pixel in line {
            print!("{pixel}");
        }
        println!();
    }
}
